master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,246 @@
+                                     mergeProperties):
+
+                # If they match, merge against this buildrequest
+                return otherBrid
+
+        # If we can't find any match, return None
+        return None
+
+    def _getSameBuildersInChain(self, startbrid, builderName):
+        """
+        :param str startbrid:
+        :param str builderName:
+        :return tuple(str,str):
+            (buildsetid, brid) for all buildrequests in the same chain (`startbrid`)
+            with the same `builderName`
+        """
+        breq_tbl = self.master.db.model.buildrequests
+
+        def __getSameBuildersInChain(conn):
+            # Select buildrequests `buildsetit` and `brid`
+            # q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+
+            # For builders in the same chain
+            #     .where(breq_tbl.c.startbrid == startbrid) \
+
+            # With the same name
+            #     .where(breq_tbl.c.buildername == builderName) \
+
+            # That were not merged themselves (only merge against one / main target)
+            #     .where(breq_tbl.c.mergebrid == None) \
+            q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+                .where(breq_tbl.c.startbrid == startbrid) \
+                .where(breq_tbl.c.buildername == builderName) \
+                .where(breq_tbl.c.mergebrid == None)
+            res = conn.execute(q)
+            return res.fetchall() or []
+
+        return self.master.db.pool.do(__getSameBuildersInChain)
+
+    def _propertiesMatch(self, properties, otherProperties, mergeProperties):
+        """
+        :param dict(str,str) properties:
+        :param dict(str,str) otherProperties:
+        :param list(str) mergeProperties:
+            List of properties that must match
+        :return bool:
+            True if `properties` and `otherProperties` match for a list of
+            `mergeProperties` to be checked
+        """
+        if 'selected_slave' in otherProperties:
+            return False
+
+        for propertyName in self.BASE_MERGE_PROPERTIES + mergeProperties:
+            if properties.get(propertyName, None) != otherProperties.get(
+                    propertyName, None):
+                return False
+
+        return True
+
+    def _getBuildsetsProperties(self, buildsetids):
+        properties = {}
+
+        # Fill in properties from cache
+        for buildsetid in buildsetids:
+            if buildsetid in self.properties_cache.cache:
+                properties[buildsetid] = self.properties_cache.get(buildsetid)
+
+        # Read missing properties in a single query
+        missing_buildsetids = set(buildsetids).difference(properties.keys())
+        if missing_buildsetids:
+            prop_tbl = self.master.db.model.buildset_properties
+
+            def __getBuildsetsProperties(conn):
+                q = sa.select(prop_tbl.c.buildsetid, prop_tbl.c.property_name, prop_tbl.c.property_value) \
```

Comment:
Overall db functions are inside buildbot.db, I think it would be best to keep that kind organization

Reply:
Will do